### DISCUSSION TITLE
Issues Cleanup

### DISCUSSION THREAD

**Discussion by yugaego** [company-mode/company-mode#1167] [General]
Hi @dgutov ,

I've got a couple of the related questions gathered with time.

1.
Company Issues section is sufficiently nicely maintained, but still,
would you like to incorporate [staling](https://github.com/apps/stale) workflow:
- exempting issues labeled `wishlist`
- with the main goal to ping participants with time to either close an issue or react to it somehow.

[Setup example](https://github.com/bbatsov/projectile/blob/master/.github/stale.yml)

2.
If mentioning a possibility to transfer some Issues to the Discussions QA section you also had on your mind doing this for the **existing** Issues, then how (if) could it be helped?
   
3.
This point I'd say more of a suggestion to be expressed.
When I was skimming through the list of the `Wishlist` issues (and looking at the open PR requests), sometimes it was unclear (especially for the earlier ones), if they're still on the agenda. And *sometimes* a potential difficulty of the implementation was unclear too. This is probably not *that* difficult to discuss per issue. But if it'd be convenient at some stage to do some cleanup here too, that would be great I think. 

--
YE

**Comment by dgutov** [comment]
1. Not very fond of that one (feels not very polite toward the reporters). And since we don't receive too many issues overall (or have a significant number hanging on some extra info provided by the user), we won't benefit from its main upside here.

2. I've been meaning to do it myself, but would you like to help? You should receive an invite with the abilities to manage issues and (hopefully) convert them to discussions. I suggest to start with the tag `question`, including those that are already closed, and see whether a question is sufficiently general (interesting to a number of users > 1) and it has one main answer or a brief discussion that can be "hidden" without losing anything important.

**Reply by dgutov** [reply]
3. I should really revisit those, and I do try, from time to time. All I can suggest here for now, though, is ask questions in specific issues.

**Reply by yugaego** [reply]
> Not very fond of that one (feels not very polite toward the reporters)

I used to feel the same. After considering for some time I found it still could be useful if done properly (f.i. better auto-replies reminding the participants an Issue needs to be acted upon, e.g. closed after response was given).
Though yes, as was already mentioned, **I surely agree it's _not_ the must-have in the case of Company**.

> I've been meaning to do it myself, but would you like to help? 

Yup, that's why I asked. Thank's for the invitation. Yes, now I see 'Convert to discussion' link beside issues.

> I suggest to start with the tag `question`, including those that are already closed, and see whether a question is sufficiently general (interesting to a number of users > 1) and it has one main answer or a brief discussion that can be "hidden" without losing anything important.

Okay, sounds good.

> ask questions in specific issues.

Yeah, that was the fallback plan.

Thanks!

**Reply by yugaego** [reply]
> 3. Wishlist issues

BTW, probably this might become handy (at some stage): just recently I discovered issues [labeled `good first issue`](https://github.blog/2020-01-22-browse-good-first-issues-to-start-contributing-to-open-source/) are handled specifically by GitHub.

**Reply by dgutov** [reply]
>  just recently I discovered issues labeled good first issue are handled specifically by GitHub

This relates to point 3 as well, I guess. :sweat_smile:

**Reply by yugaego** [reply]
Ah, the first one I had transferred missed the point of being "popular" enough (user > 1), even though it seemed to be general.
And I've seen a couple of similar ones: they *almost* matched the criteria.
Maybe I could just add a `transfer?` label to such Issues for you to review, until I've got a better grasp of what exactly would serve Company needs the best.

**Reply by dgutov** [reply]
> Ah, the first one I had transferred missed the point of being "popular" enough (user > 1), even though it seemed to be general.

This one? #1168. Seems general enough. I just meant that it should be something more than 1 user would want to know. And that one surely hits the mark. I just had to move it to Q&A, though.

Feel free to add that label to the ones you're not sure about, of course.

**Reply by yugaego** [reply]
Got it, thanks 👍🏻

**Comment by yugaego** [comment]
@dgutov , is the [following](https://github.com/company-mode/company-mode/issues/959#issuecomment-595818130) regarding `company-bbdb` (or another backend) still true:

>  If I was making that decision now, I would also ask the submitter to create a completion-at-point function instead. If it's a good fit, of course. People are welcome to do that, and when that happens I'd happily remove that backend.
?

If yes, would it be probably a good idea to either create a `wishlist` Issue(s) or at least create new `Idea` topic(s) here, in Discussions?

**Reply by dgutov** [reply]
Not not sure which one to pick, but that should make sense.

It's a bit weird, though: to file a task for someone to extract a piece of code and move it to another project.

**Reply by yugaego** [reply]
> It's a bit weird, though: to file a task for someone to extract a piece of code and move it to another project.

Yeah, I understand. I'm fine with any choice, including any of:
- Doing nothing more
  - (for the foreseeable future or at the moment).
  - At least, it's already being discussed in this thread.
- Adding this topic to the `Ideas` section
  -  (since it seems to fit a project's "design/style idealistic notion" / "design goals" category).
  - Might be *Roadmap* is the correct term I've been looking for on the previous line.
- Creating an Issue with a particular assignee
  - if this notion truly meets your vision for the project
  - (f.i. I might be interested in working on it with time).

**Reply by dgutov** [reply]
Very good, I've filed the issue: https://github.com/company-mode/company-mode/issues/1176

You're welcome to self-assign whenever you find the time and inspiration. :+1:

**Reply by yugaego** [reply]
> You're welcome to self-assign whenever you find the time and inspiration.

Yup, sounds good. There're already a couple of wished issues I'm interested to look into, but postponing to after I'm finished with the Elisp documentation (which takes longer than I had expected (what doesn't 😄).

**Reply by yugaego** [reply]
Issue #923 seems to be related to the discussion above (#959).

Should the guidelines from https://github.com/company-mode/company-mode/issues/923#issuecomment-544263928 (and other comments from that thread) be documented somewhere?
(*Was that maybe the meaning of the `wishlist` label?*)

**Maybe** it's worth adding [Contributing file](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/setting-guidelines-for-repository-contributors) and link to it from the `Discussions` section and the `Homepage`?

That file could contain two sections describing how to:
- Contribute fixes/improvements to the existing code-base.
- Integrate/add a backend.

Another approach could be posting instructions on how to integrate/add a backend to the `Discussions` (QA) section.
Though in this case auto-transfer from the issue doesn't seem to me to be a good idea.

What do you think?

**Reply by dgutov** [reply]
> Issue #923 seems to be related to the discussion above (#959).

I think those are different things: one is about third-party packages modifying the variable `company-backends` somehow: either during their initialization, or by asking their users, in the docs, to add some invocations to the init script. The backend definition (or `completion-at-point-functions` element) lives outside of this package then. And the other is about adding to the list of built-in backends.

> (Was that maybe the meaning of the wishlist label?)

Pretty much. About documenting those guidelines, as well as doing a survey of the current practice in the popular packages. To make sure we either document the existing practice, or (if it's different) there is at least a good chance to have those new guidelines followed.

> Maybe it's worth adding Contributing file and link to it from the Discussions section and the Homepage?

I think adding a Contributing file can be a good idea, but what would it contain?

> Contribute fixes/improvements to the existing code-base.

This section will be either empty or... some common sense stuff? Like: propose an idea, then see the reaction, implement, maintain existing coding style, then and listen to the reviews. Or take on an existing issue and then do the rest of the steps. And don't forget to do the copyright assignment thing.

> Integrate/add a backend.

And this would probably say that "we don't, in general, accept new backends". Especially niche ones. Though there can be exceptions.

> Another approach could be posting instructions on how to integrate/add a backend to the Discussions (QA) section.

Probably not. I'm not sure how that would look, or what route the user would take to end up on that page.

**Reply by yugaego** [reply]
>  some common sense stuff? Like: propose an idea, then see the reaction, implement, maintain existing coding style, then and listen to the reviews. Or take on an existing issue and then do the rest of the steps. And don't forget to do the copyright assignment thing.

That's right. It can be concise but informative.
(arguably, all these common files/texts are helpful for newbies and serve as a kind of "visit cards" for the rest). 
And yes, I thought of something in the spirit of what you'd suggested: 
- Pull requests are welcome and wanted.
- `Wishlist` issues are for everyone to take on (fun fact: 33 are open and 33 are closed now).
- Copyright assignment description from the homepage would be basically duplicated or moved to this file.

> This would probably say that "we don't, in general, accept new backends". Especially niche ones. Though there can be exceptions.  

Haha, but yes. And I also meant text based on the comments from #923, (one example is https://github.com/company-mode/company-mode/issues/923#issuecomment-544263928), so closer to what you'd described in the previous comment here (with the most difficult part to me seemingly being  "as well as doing a survey of the current practice in the popular packages.").

I may start compiling the texts a bit later, so we could see how it goes.

**Reply by dgutov** [reply]
> And yes, I thought of something in the spirit of what you'd suggested:

Sounds good.

> Wishlist issues are for everyone to take on (fun fact: 33 are open and 33 are closed now).

With the caveat that the trivial ones I usually fix on the spot, so the ones remaining often need some redesign (minor or not), or a re-imagining of a feature.

Fun fact indeed. :-)

> And I also meant text based on the comments from #923, (one example is #923 (comment)), so closer to what you'd described in the previous comment here

Sure. Maybe even add a more detailed guideline for how to choose which mechanism to pick. So this is a section about third-party packages? Not how to "add a backend" to company-mode?

> with the most difficult part to me seemingly being "as well as doing a survey of the current practice in the popular packages."

It's a social thing, which can be daunting sometimes. On the bright side, there are only a handful of really popular company backends these days (including packages using c-a-p-f functions), so we can talk to them and prioritize that info.

> I may start compiling the texts a bit later, so we could see how it goes.

Thanks!

**Reply by yugaego** [reply]
> About documenting those guidelines, as well as doing a survey of the current practice in the popular packages. To make sure we either document the existing practice, or (if it's different) there is at least a good chance to have those new guidelines followed.

I reviewed the third-party packages listed on the Wiki page, and it seems to me the situation is quite fine.
A couple of the packages define a CAPF function, roughly (I didn't really count) 40%  follow a local variable approach and the rest of the packages adjust `company-backends` globally checking for the major mode.

In other words, I saw an overall accordance (though not always perfect in terms of documenting the behavior) with the guidelines. Anyway, I don't feel a need to do anything else in this regard at this stage, so I'm for closing the ticket unless you have a reason to keep it open.

**Comment by yugaego** [comment]
Hi Dmitry, please take a look at the Issue #511 .
It seems to be a good candidate to be converted to a QA (and has recent links to it), but I'm a bit hesitant:

Are you still positive there's no need to extract a defcustom suggested by your first answer?

Could `"[[:upper:]]+"` be enough for that particular cases-sorting question?
(Regexp documentation for those quotes is not exactly clear or I may be missing a use case.)

**Reply by dgutov** [reply]
Hi Yuzhana,

It can be made into a built-in transformer function, and mentioned in the `company-transformers` `:type` form.

Do you think it's going to be used by >1 user, though? I've never seen anybody else ask for this.

> Could "[[:upper:]]+" be enough for that particular cases-sorting question?

I took the question literally and wrote a regexp what matches only fully upper-case strings. Whether it's a good idea, probably depends on what the other users would want to use it for.

> (Regexp documentation for those quotes is not exactly clear or I may be missing a use case.)

`\\'` means "end of the string/buffer" and the other one means "beginning". Did you mean those quotes?

**Reply by yugaego** [reply]
> Do you think it's going to be used by >1 user, though?

I actually don't know.
The thread interested me mostly as a potential QA to present an example (use-case with the screenshot) on how to write and use `company-transformers`.
But there're also `defcustom` extraction example and two links from other repos made in 2020 discussing candidates sorting with which I wasn't sure what to do.

Now I can rephrase the question:
*Does it seem fine to you to convert "Sort uppercase completions to the bottom" to QA with this [answer](https://github.com/company-mode/company-mode/issues/511#issuecomment-219016047)?*

> Did you mean those quotes?

Yep. Your clarifications explained their meaning, thanks! 😉.

**Reply by dgutov** [reply]
> Does it seem fine to you to convert "Sort uppercase completions to the bottom" to QA with this answer?

Yes, that seems like the best choice: both incoming links to that issue are about sorting, but the sorting rules they want are all different.

Thanks for the help!

**Comment by yugaego** [comment]
Hey Dmitry, FYI: I've finished reviewing the `question` issues 🥳 .
That was an interesting journey, even unexpectedly relaxing.
One may say there's potentially a whole new field of *recreational triaging* 😸 
See you later then.

**Reply by dgutov** [reply]
A little bit like gardening, I guess. :-)

I also like that part of open source.

**Comment by yugaego** [comment]
Hey Dmitry, please take a look, here's [one more issue](https://github.com/company-mode/company-mode/issues/738) that probably should be converted to a discussion. [I myself don't see the convert/transfer button anymore.]

**Reply by yugaego** [reply]
And [one more](https://github.com/company-mode/company-mode/issues/1066), for this part
> Q "Running company-complete with the point at the beginning of a line, when in sh-mode, triggers a long list of completion options. Is this by design? "
 > A "if you ask for completion on an empty prefix, it will give you whatever completions it has."

I found these Q/A looking for an answer to [this SE question](https://stackoverflow.com/questions/69439076/how-to-prevent-automatic-auto-completion-window-pop-up-after-class-member-access?noredirect=1#comment122760788_69439076) (and digging into `Company` internals more).

Only now I understood the issue was most probably caused by the empty prefix value rather than by some backend behavior.
I've considered adding this understanding as a docstring but didn't find a proper place to put it. Any hints here?

Thank you.

**Reply by dgutov** [reply]
> [I myself don't see the convert/transfer button anymore.]

That's odd. :-( You still have the "Triage" role which should give access to that function. Maybe they have changed something in how permissions work.

Thanks, I've converted it to a Discussion.

**Reply by dgutov** [reply]
> And one more, for this part

It probably includes some interesting insight, but the headline is not helpful. So when converting it to a question someone will need to create a one-liner question that summarizes the issue body. Right?

> I found these Q/A looking for an answer to this SE question (and digging into Company internals more).

This one is a little different: the question is why the completion popup appears automatically. Not why it offers so many completions.

And the line you linked to (https://github.com/jorgenschaefer/elpy/blob/325f6ed000a115daf419e8d614fdce22aa95a78e/elpy.el#L3014) should be the one responsible, because it overrides the "effective prefix length", this triggering idle completion (which otherwise is triggered when the prefix's length reaches the value of `company-minimum-prefix-length`).

**Reply by dgutov** [reply]
However, when I set `company-idle-delay` to `nil`, I don't get any idle completion, whether it's during normal typing or after `.` in `c-mode` with `company-clang` (which also uses `company-grab-symbol-cons`).

So I'm not 100% sure what's going on there.

**Reply by yugaego** [reply]
> That's odd. :-( You still have the "Triage" role

Yeah. There were reports about [similar issues](https://github.com/github/feedback/discussions/2924#discussioncomment-254592
) months ago.
I decided not to bother since the main part of the transfer had been done.

> but the headline is not helpful.

Yes, I thought so.
How about copying the phrase from the question
"Is there a way of forcing company-complete to ignore the empty string?".
It contains useful keywords that a `Company` newbie would probably use.
And since the answer contains the term ("empty") "prefix", the search in Discussions should work fine for a more acknowledged person too.

> the question is why the completion popup appears automatically.

Thank you! Your answers have clarified my understanding.

**Reply by yugaego** [reply]
And I still have access to the opposite action "Create issue from discussion", BTW.

**Reply by dgutov** [reply]
> There were reports about similar issues months ago.

Sounds different: I have the button available, after all. And you don't. And the report is about the feature being disabled (at the beginning of this year).

> And I still have access to the opposite action "Create issue from discussion", BTW.

:confounded:

**Reply by dgutov** [reply]
> How about copying the phrase from the question
"Is there a way of forcing company-complete to ignore the empty string?".

Okay, but then the answer would be something different rather than what I said (and the author accepted).

The answer would need to be somewhere along the lines of adding an around-advice on `sh-completion-at-point-function`, or adding a new element to `company-transformers` which would look at `company-prefix`, see it's an empty string, and return nil.

Or changing the default value of `company-candidates-predicate` to do the same.

Neither approach seems universally useful/TRT, to be clear, but it's what can be done.

**Reply by yugaego** [reply]
> but then the answer would be something different rather than what I said (and the author accepted).
> Neither approach seems universally useful/TRT, to be clear, but it's what can be done.

Then probably the last message here can be taken and suggested as an optional, not "universally useful ... what can be done" approach.
Then:
- either the existing answer could be edited.
- or one more answer could be added, with the issue transferred to the "General" category, so there won't be need to choose an answer.

Or do nothing.
Or, as I had mentioned, maybe find a good place (I didn't/couldn't) for a docstring saying "if you ask for completion on an empty prefix, it will give you whatever completions it has.".

**Reply by dgutov** [reply]
Very good. Thanks, added. :-)

**Reply by yugaego** [reply]
Looks great. Thank you.
